Stripe/Payments/STPPaymentManager.m

```diff
@@ -0,0 +1,406 @@
+
+                    [_apiClient authenticate3DS2:authRequestParams
+                                sourceIdentifier:authenticationAction.useStripeSDK.threeDS2SourceID
+                                      completion:^(STP3DS2AuthenticateResponse * _Nullable authenticateResponse, NSError * _Nullable error) {
+                                          if (authenticateResponse == nil) {
+                                              completion(STPPaymentManagerActionStatusFailed, self->_currentAction.paymentIntent, error);
+                                          } else {
+                                              STDSChallengeParameters *challengeParameters = [[STDSChallengeParameters alloc] initWithAuthenticationResponse:authenticateResponse.authenticationResponse];
+
+                                              [transaction doChallengeWithViewController:[self->_currentAction.authenticationContext authenticationPresentingViewController]
+                                                                     challengeParameters:challengeParameters
+                                                                 challengeStatusReceiver:self
+                                                                                 timeout:self->_currentAction.threeDSCustomizationSettings.authenticationTimeout];
+                                          }
+                                      }];
+                }
+                    break;
+            }
+            break;
+    }
+}
+
+- (void)_retrieveAndCheckPaymentIntentForCurrentAction {
+    [_currentAction.apiClient retrievePaymentIntentWithClientSecret:_currentAction.paymentIntent.clientSecret
+                                           completion:^(STPPaymentIntent * _Nullable paymentIntent, NSError * _Nullable error) {
+                                               if (error != nil) {
+                                                   self->_currentAction.completion(STPPaymentManagerActionStatusFailed, paymentIntent, error);
+                                               } else {
+                                                   self->_currentAction.paymentIntent = paymentIntent;
+                                                   [self _handleNextActionIfNeededAttemptAuthentication:NO];
+                                               }
+                                           }];
+}
+
+- (void)_handleWillForegroundNotification {
+    [[NSNotificationCenter defaultCenter] removeObserver:self name:UIApplicationWillEnterForegroundNotification object:nil];
+    [self _retrieveAndCheckPaymentIntentForCurrentAction];
+}
+
+#pragma mark - SFSafariViewControllerDelegate
+
+- (void)safariViewControllerDidFinish:(SFSafariViewController * __unused)controller {
+    [[STPURLCallbackHandler shared] unregisterListener:self];
+    [self _retrieveAndCheckPaymentIntentForCurrentAction];
+}
+
+#pragma mark - STPURLCallbackListener
+
+- (BOOL)handleURLCallback:(NSURL * __unused)url {
```

Comment:
Now that we're not pushing users out to Safari, does this path still make sense?   Or, what happens if the `SFSafariViewController` redirects to the app?

Reply:
Yeah we still need this for the case where deep-linking actually takes us to another app

Reply:
er, actually let me test